packages/components/nodes/vectorstores/Upstash/Upstash.ts

```diff
@@ -0,0 +1,181 @@
+		]
+		this.outputs = [
+			{
+				label: 'Upstash Retriever',
+				name: 'retriever',
+				baseClasses: this.baseClasses
+			},
+			{
+				label: 'Upstash Vector Store',
+				name: 'vectorStore',
+				baseClasses: [this.type, ...getBaseClasses(UpstashVectorStore)]
+			}
+		]
+	}
+
+	//@ts-ignore
+	vectorStoreMethods = {
+		async upsert(nodeData: INodeData, options: ICommonObject): Promise<Partial<IndexingResult>> {
+
+			const docs = nodeData.inputs?.document as Document[]
+			const embeddings = nodeData.inputs?.embeddings as Embeddings
+			const recordManager = nodeData.inputs?.recordManager
+
+			const credentialData = await getCredentialData(nodeData.credential ?? '', options)
+			const UPSTASH_VECTOR_REST_URL = getCredentialParam('UPSTASH_VECTOR_REST_URL', credentialData, nodeData)
+			const UPSTASH_VECTOR_REST_TOKEN = getCredentialParam('UPSTASH_VECTOR_REST_TOKEN', credentialData, nodeData)
+
+			const upstashIndex = new UpstashIndex({
+				url: UPSTASH_VECTOR_REST_URL,
+				token: UPSTASH_VECTOR_REST_TOKEN
+			})
+
+			const flattenDocs = docs && docs.length ? flatten(docs) : []
+			const finalDocs = []
+			for (let i = 0; i < flattenDocs.length; i += 1) {
+				if (flattenDocs[i] && flattenDocs[i].pageContent) {
+					finalDocs.push(new Document(flattenDocs[i]))
+				}
+			}
+
+			const obj = {
+				index: upstashIndex
+			}
+
+			try {
+				if (recordManager) {
+					const vectorStore = await UpstashVectorStore.fromExistingIndex(embeddings, obj)
+					await recordManager.createSchema()
+					const res = await index({
+						docsSource: finalDocs,
+						recordManager,
+						vectorStore,
+						options: {
+							cleanup: recordManager?.cleanup,
+							sourceIdKey: recordManager?.sourceIdKey ?? 'source',
+							vectorStoreName: UPSTASH_VECTOR_REST_URL
+						}
+					})
+
+					return res
+				} else {
+					await UpstashVectorStore.fromDocuments(finalDocs, embeddings, obj)
+					return { numAdded: finalDocs.length, addedDocs: finalDocs }
+				}
+			} catch (e) {
+				throw new Error(e)
+			}
+		}
+	}
+
+	async init(nodeData: INodeData, _: string, options: ICommonObject): Promise<any> {
+		const upstashMetadataFilter = nodeData.inputs?.upstashMetadataFilter
+		const embeddings = nodeData.inputs?.embeddings as Embeddings
+
+		const credentialData = await getCredentialData(nodeData.credential ?? '', options)
+		const UPSTASH_VECTOR_REST_URL = getCredentialParam('UPSTASH_VECTOR_REST_URL', credentialData, nodeData)
+		const UPSTASH_VECTOR_REST_TOKEN = getCredentialParam('UPSTASH_VECTOR_REST_TOKEN', credentialData, nodeData)
+
+		const upstashIndex = new UpstashIndex({
+			url: UPSTASH_VECTOR_REST_URL,
+			token: UPSTASH_VECTOR_REST_TOKEN
+		})
+
+		const obj: UpstashVectorStoreParams = {
+			index: upstashIndex
+		}
+
+		if (upstashMetadataFilter) {
+			obj.filter = upstashMetadataFilter
+		}
+
+		const vectorStore = await UpstashVectorStore.fromExistingIndex(embeddings, obj)
+
+		return resolveVectorStoreOrRetriever(nodeData, vectorStore)
```

Comment:
can we also add `obj.filter`:

`return resolveVectorStoreOrRetriever(nodeData, vectorStore, obj.filter)`